Update install_tracked for newer versions of Moo, fix roles

zmughal opened the pull request:
Use `_install_tracked` from `Moo::_Utils` directly
    
This allows for roles to work again with newer versions of Moo as
`_install_tracked` is shared as an API for both `Moo` and `Moo::Role`.
See <https://github.com/moose/Moo/commit/a72639a9b6259e487e0cf496cf9d8d097e62ab39>
for more info. Available since Moo-2.003006.

zmughal committed:
Add TODO test for using with Moo::Role

zmughal committed:
Use _install_tracked from Moo::_Utils directly

This allows for roles to work again with newer versions of Moo as
`_install_tracked` is shared as an API for both `Moo` and `Moo::Role`.
See <https://github.com/moose/Moo/commit/a72639a9b6259e487e0cf496cf9d8d097e62ab39>
for more info. Available since Moo-2.003006.

The approach here is to check the Moo version and use the API directly
if available, otherwise use older installer functions.

robrwo commented:
Looks ok to me

zmughal commented:
Just did a small whitespace fix. Should be ready to go now.

vanHoesel commented:
does this jump the Moo dependency too ? if so, it would be nice to have it declared in the `cpanfile`.

or, if we do not want to bump the Moo dependency, is there a way to be backward compatible of versions before 2.003006 ?

robrwo commented:
If the Moo version dependency is to be changed, do it in the module itself, or in `dist.ini`. Dist::Zilla will update the cpanfile automatically.

(The cpanfile is so that devs can install the dependencies locally and use Carton.)

robrwo commented:
Actually yes, this needs the Moo version updated. So update Should.pm to have the minimum version.

zmughal commented on lib/MooX/Should.pm:
```diff
@@ -2,8 +2,8 @@ package MooX::Should;
 
-use Moo       ();
-use Moo::Role ();
+use Moo 2.003006 ();
```
I have placed the version here. Is this fine? When I run `dzil build`, it updates the `cpanfile` with this version as expected.

robrwo commented:
Hm, even better yet, modify this so that it checks the Moo version. If the version is newer, then it uses the newer API, otherwise it falls back.

zmughal commented:
>  If the version is newer, then it uses the newer API, otherwise it falls back.

I'll update with that in a few.

zmughal commented:
I think the module did not work with roles in the first place. It appears that Moo::Role::install_tracked never existed at least for version of `Moo` that was stable for the initial release of `MooX::Should`:

```shell
$ ( cpanm -nq Moo@2.003004; prove -lvr t );
Successfully installed Moo-2.003004
1 distribution installed
[...]
t/03-role.t ....
not ok 1 # TODO Not able to use role with MooX::Should: dies due to missing installer function
#   Failed (TODO) test at t/03-role.t line 15.
# died: Undefined subroutine &Moo::Role::install_tracked called at /home/zaki/sw_projects/zmughal/MooX-Should/MooX-Should/lib/MooX/Should.pm line 34.
# BEGIN failed--compilation aborted at t/lib/TestRole.pm line 4.
# Compilation failed in require at /home/zaki/perl5/perlbrew/perls/perl-5.26.1/lib/site_perl/5.26.1/Module/Runtime.pm line 314.
1..1
ok
```

So the simplest fix is to just change that and it will work for all versions of `Moo`.

I'll still push code that checks the version.

zmughal committed:
Fix name of installer function for Moo::Role

Previous versions of Moo::Role (for example under Moo-2.003004) had
`Moo::Role::_install_tracked` and not `Moo::Role::install_tracked`.

zmughal commented:
```shell
-(use-moo-utils=)-$ ( cpanm -nq Moo@2.003004; prove -lvr t ); ( cpanm -nq Moo@2.003006; prove -lvr t ); ( cpanm -nq Moo; prove -lvr t )
Successfully installed Moo-2.003004 (downgraded from 2.004004)
1 distribution installed
t/01-strict.t ..
ok 1 - isa
ok 2 - should
ok 3 - isa
ok 4 - should
1..4
ok
t/02-lax.t .....
ok 1 - isa
ok 2 - should
ok 3 - isa
ok 4 - should
1..4
ok
t/03-role.t ....
ok 1
1..1
ok
All tests successful.
Files=3, Tests=9,  0 wallclock secs ( 0.02 usr  0.00 sys +  0.42 cusr  0.07 csys =  0.51 CPU)
Result: PASS
Successfully installed Moo-2.003006 (upgraded from 2.003004)
1 distribution installed
t/01-strict.t ..
ok 1 - isa
ok 2 - should
ok 3 - isa
ok 4 - should
1..4
ok
t/02-lax.t .....
ok 1 - isa
ok 2 - should
ok 3 - isa
ok 4 - should
1..4
ok
t/03-role.t ....
ok 1
1..1
ok
All tests successful.
Files=3, Tests=9,  0 wallclock secs ( 0.02 usr  0.01 sys +  0.46 cusr  0.07 csys =  0.56 CPU)
Result: PASS
Successfully installed Moo-2.004004 (upgraded from 2.003006)
1 distribution installed
t/01-strict.t ..
ok 1 - isa
ok 2 - should
ok 3 - isa
ok 4 - should
1..4
ok
t/02-lax.t .....
ok 1 - isa
ok 2 - should
ok 3 - isa
ok 4 - should
1..4
ok
t/03-role.t ....
ok 1
1..1
ok
All tests successful.
Files=3, Tests=9,  0 wallclock secs ( 0.03 usr  0.00 sys +  0.42 cusr  0.06 csys =  0.51 CPU)
Result: PASS
```

robrwo commented on lib/MooX/Should.pm:
```diff
@@ -2,9 +2,20 @@ package MooX::Should;
+
 use Devel::StrictMode;
 
 our $VERSION = 'v0.1.3';
```
If we're going to use `version`, then we might as well use
```perl
our $VERSION = version->declare('v0.1.3');
```

zmughal replied:
Done!

zmughal committed:
Use version object for $VERSION

vanHoesel commented:
hey folks, you're updating faster then I can review

vanHoesel reviewed (approved):
I'm okay with these, looks all good to me, love the backwards compatibility check.

robrwo commented:
I've updated the Changes and README. Thanks.
